File: 兴趣班入门百练讲义/NQ100/NQ052/NQ052_improved.cpp
#include <iostream>
#include <algorithm>
using namespace std;

void quickSort(long long nums[], int left, int right) {
    if (left >= right) return;
    long long pivot = nums[left];
    int i = left - 1, j = right + 1;
    while (i < j) {
        do i++; while (nums[i] < pivot);
        do j--; while (nums[j] > pivot);
        if (i < j) swap(nums[i], nums[j]);
    }
    quickSort(nums, left, j);
    quickSort(nums, j + 1, right);
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n;
    cin >> n;
    long long* nums = new long long[n];
    for (int i = 0; i < n; i++) cin >> nums[i];
    quickSort(nums, 0, n - 1);
    for (int i = 0; i < n; i++) cout << nums[i] << " ";
    cout << endl;
    delete[] nums;
    return 0;
}

File: source/algorithm/NQ052/NQ052_improved.py
import sys

def quick_sort(nums, left, right):
    if left >= right:
        return
    pivot = nums[left]
    i, j = left - 1, right + 1
    while i < j:
        i += 1
        while nums[i] < pivot:
            i += 1
        j -= 1
        while nums[j] > pivot:
            j -= 1
        if i < j:
            nums[i], nums[j] = nums[j], nums[i]
    quick_sort(nums, left, j)
    quick_sort(nums, j + 1, right)

def main():
    input = sys.stdin.read().split()
    n = int(input[0])
    nums = list(map(int, input[1:n+1]))
    quick_sort(nums, 0, n - 1)
    print(' '.join(map(str, nums)))

if __name__ == "__main__":
    main()